packages/component-base/src/slot-controller.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+export class SlotController {
+  constructor(host, slotName, slotFactory, slotInitializer) {
+    this.host = host;
+    this.slotName = slotName;
+    this.slotFactory = slotFactory;
+    this.slotInitializer = slotInitializer;
+  }
+
+  hostConnected() {
+    if (!this.initialized) {
+      const { host, slotName, slotFactory, slotInitializer } = this;
+
+      const slotted = this.getSlotChild(slotName);
+
+      if (!slotted) {
+        // Slot factory is optional, some slots don't have default content.
+        if (slotFactory) {
+          const slotContent = slotFactory(host);
+          if (slotContent instanceof Element) {
+            if (slotName !== '') {
+              slotContent.setAttribute('slot', slotName);
+            }
+            host.appendChild(slotContent);
+            this.node = slotContent;
+          }
+        }
+      } else {
+        this.node = slotted;
+      }
+
+      // Don't try to bind `this` to initializer (normally it's arrow function).
+      // Instead, pass the host as a first argument to access component's state.
+      if (slotInitializer) {
+        slotInitializer(host, this.node);
+      }
+
+      this.initialized = true;
+    }
+  }
+
+  /**
+   * Get a reference to the node managed by the controller.
+   * @param {string} slotName
```

Comment:
The `@return` JSDoc annotation is missing.

Reply:
Fixed.